bench/macro/lsm-tree-bench-unions.hs

```diff
@@ -0,0 +1,629 @@
+    , check               :: !Bool
+    }
+  deriving stock Show
+
+data Cmd
+    -- | Setup benchmark: generate initial LSM trees etc.
+    = CmdSetup
+
+    -- | Make a dry run, measure the overhead.
+    | CmdCollisions
+
+    -- | Run the actual benchmark
+    | CmdRun RunOpts
+  deriving stock Show
+
+-------------------------------------------------------------------------------
+-- command line interface
+-------------------------------------------------------------------------------
+
+globalOptsP :: O.Parser GlobalOpts
+globalOptsP = pure GlobalOpts
+    <*> O.option O.str (O.long "bench-dir" <> O.value "_bench_unions" <> O.showDefault <> O.help "Benchmark directory to put files in")
+    <*> O.option (greaterThanOneParser "number of tables") (O.long "table-count" <> O.value 10 <> O.showDefault <> O.help "Number of tables to benchmark")
+    <*> O.option (greaterThanOneParser "size of tables") (O.long "initial-size" <> O.value 1_000_000 <> O.showDefault <> O.help "Initial size of each table")
+    <*> O.option (Just <$> O.auto) (O.long "seed" <> O.value Nothing <> O.showDefault <> O.help "Random seed, uses system entropy when not explicitly specified")
+  where
+    greaterThanOneParser str =
+        let validator v
+                | v >= 2 = Right v
+                | otherwise = Left $ unwords
+                    [ "Negative number for", str <> ":", show v ]
+        in  O.eitherReader $ \msg -> Read.readEither msg >>= validator
+
+cmdP :: O.Parser Cmd
+cmdP = O.subparser $ mconcat
+    [ O.command "setup" $ O.info
+        (CmdSetup <$ O.helper)
+        (O.progDesc "Setup benchmark by generating required tables")
+    , O.command "collisions" $ O.info
+        (CmdCollisions <$ O.helper)
+        (O.progDesc "Collision analysis, compute theoretical and actual shared keys between tables")
+
+    , O.command "run" $ O.info
+        (CmdRun <$> runOptsP <**> O.helper)
+        (O.progDesc "Proper run, measuring performance and generating a benchmark report")
+    ]
+
+runOptsP :: O.Parser RunOpts
+runOptsP = pure RunOpts
+    <*> O.option O.auto (O.long "batch-count" <> O.value 200 <> O.showDefault <> O.help "Batch count")
+    <*> O.option O.auto (O.long "batch-size" <> O.value 256 <> O.showDefault <> O.help "Batch size")
+    <*> O.switch (O.long "check" <> O.help "Check generated key distribution")
+
+-------------------------------------------------------------------------------
+-- measurements
+-------------------------------------------------------------------------------
+
+timed :: IO a -> IO (a, Integer)
+timed action = do
+    performMajorGC
+    t1 <- Clock.getTime Clock.Monotonic
+    x  <- action
+    t2 <- Clock.getTime Clock.Monotonic
+    performMajorGC
+    let !ms = Clock.toNanoSecs (Clock.diffTimeSpec t2 t1) `div` 1_000_000
+    return (x, ms)
+
+-- | The 'Integer' is the number of /milliseconds/ elapsed.
+timed_ :: IO () -> IO Integer
+timed_ action = do
+    ((), t) <- timed action
+    pure t
```

Comment:
I noticed elsewhere that you measure timing for *each* batch lookup, in which case milliseconds are probably too coarse a unit.

Reply:
I changed the coarseness to microseconds, rather than aggregate all the samples into a single measurement.